### PR TITLE
lib.ipsec.esp: second iteration [wip]

### PR DESCRIPTION
**Pull request by eugeneia**
The code in `lib.ipsec.esp` is in a “proof of concept” stage but lacks tests, in the next few days I will hopefully be able to address the following issues:
- Obtain reference PCAP dumps from a known ESP flow (both encrypted and decrypted), e.g.: `aes128-gcm12` IPv6 transport mode, with known key/salt/SPI values
- Add unit tests that confirm our implementation is able to truthfully produce and decrypt above PCAP dumps
- Add a benchmark to confirm performance to match the Intel AES-NI whitepaper

The current implementation has the following known bugs that need to be addressed:
- Transmit truncated 32bit sequence number (don't implement anti-replay on receiver side for now), see https://tools.ietf.org/html/rfc4303#section-3.3.3. As of now `lib.ipsec.esp` sends a 64bit sequence number, of which the first four octets are stored in the ESP header and the other four are sent as the initialization vector. Sequence number verification is currently broken and should be removed.
- Transmit independent 64bit IV in front of ciphertext, see https://tools.ietf.org/html/rfc4106#section-3. As of now the 64bit sequence number is wrongly used as the IV.
- Correctly implement minimal padding (ciphertext aligned to 4 byte boundary, see
  https://tools.ietf.org/html/rfc4303#section-2.4 and https://tools.ietf.org/html/rfc4106#section-7).

Cc @hb9cwp to keep him in the loop.

### CODE REVIEW

**Comment by eugeneia** [comment]
> Obtain reference PCAP dumps from a known ESP flow (both encrypted and decrypted), e.g.: aes128-gcm12 IPv6 transport mode, with known key/salt/SPI values

@lukego  @capr I have looked into this briefly today and couldn't find a reference implementation from which we can obtain these, any pointers wold be appreciated. Linux `setkey` bundled with Ubuntu trusty tahr doesn't support `aes128-gcm-12` as far as I could tell from the Manpage. There is strongSwan but I don't think it lets you define static security associations easily.

**Comment by hb9cwp** [comment]
@eugeneia Interesting! Also, I am tempted to try and add test vectors from the paper [GCM] referenced in https://tools.ietf.org/html/rfc4106  to ESP-GCM-128...

**Comment by eugeneia** [comment]
@hb9cwp great catch! Yes we should test against those.

**Comment by eugeneia** [comment]
On a second look I noticed these are just AES test vectors, not including ESP headers etc... I think we need a full session flow of aes128-gcm12.

**Comment by eugeneia** [comment]
Correction: `lib.ipsec.esp` implements aes128-gcm16, **not** aes128-gcm12

**Comment by hb9cwp** [comment]
With the trailing "16" you probably refer to the Integrity Check Value (ICV) with the 16 Byte Authentication Tag which ESM-GCM appends to the very end of the packet? 
As per RFC4106 section 6, mandatory is 16 Byte, whereas 8 or 12 Byte tags are optional. Indeed, `aes_128_gcm.auth_size` is always 16 in the current implementation.

**Comment by hb9cwp** [comment]
A small drawing with the packet format from my test & debug sessions (for a small payload with 64 Byte, as in selftest() ):
rs22012016

**Comment by eugeneia** [comment]
@hb9cwp Yes I meant the ICV length.

**Comment by hb9cwp** [comment]
Reading RFC4106 and references reveals that implementation of padding is off by a factor of 8, e.g. `aes_128_gcm.lua` declares `o.blocksize = 128`, which indeed correspond to the blocksize of AES in bits, whereas `esp.lua` uses `o.blocksize` in Bytes, see my commit https://github.com/hb9cwp/snabbswitch/commit/274f42a7a298b5f713f5a93db3fe51a224dd0cae
In the worst case, this results in 8 times more padding than necessary (127 instead of 15 Byte), e.g. way more overhead than necessary with AES-GCM! 
Although my drawing above reflects the current implementation, I'll have to re-draw it with the correct blocksize and padding...
@eugeneia This partially addresses your work item above "Correctly implement minimal padding (ciphertext aligned to 4 byte boundary, ...)".

**Comment by hb9cwp** [comment]
@eugeneia  In terms of defensive programming, allocate one Byte more than strictly necessary for `o.pad_buf` in `esp.lua`, see my commit https://github.com/hb9cwp/snabbswitch/commit/0bd2fbface9a66e43f371c69eeafcbcf6ce4b648 

The last Byte should never be overwritten. Thus we could use it as a canary to detect buffer overflow.

**Comment by eugeneia** [comment]
@hb9cwp Yes, the padding code was completely broken and mostly served to exercise the ESP padding fields. Check out 824a665 which intends to make `lib.ipsec.esp` RFC compliant.

File: src/lib/ipsec/esp.lua [hunk 1]
```diff
@@ -76,28 +91,29 @@ local esp_v6_decrypt = {}
...
+   local esp_overhead = esp_size + esp_tail_size + gcm.iv_size + gcm.auth_size
+   o.min_size = esp_overhead + padding(o.pad_to, esp_overhead)
    return setmetatable(o, {__index=esp_v6_decrypt})
 end
 
 -- Verify sequence number.
 function esp_v6_decrypt:check_seq_no (seq_no)
-   self.seq_no = self.seq_no + 1
-   return self.seq_no <= seq_no
+   -- FIXME: Check seq_no here, see https://tools.ietf.org/html/rfc4303#page-38
```
Comment (eugeneia) [inline src/lib/ipsec/esp.lua]: Still missing: anti-replay protection.

**Comment by eugeneia** [comment]
@hb9cwp Regarding the minimum padding: the way I understood it, we need to pad to the four byte boundary. E.g. at most three bytes. See https://tools.ietf.org/html/rfc4303#section-2.4 and https://tools.ietf.org/html/rfc4106#section-3.2

ESP RFC:

> Padding also may be required, irrespective of encryption algorithm requirements, to ensure that the resulting ciphertext terminates on a 4-byte boundary.  Specifically, the Pad Length and Next Header fields must be right aligned within a 4-byte word, as illustrated in the ESP packet format figures above, to ensure that the ICV field (if present) is aligned on a 4-byte boundary.

GCM in IPsec RFC:

> Implementations that do not seek to hide the length of the plaintext SHOULD use the minimum amount of padding required, which will be less than four octets.

I have tested passing 56 byte long cleartext (e.g. not aligned to a 128bit boundary) to `aesni_gcm_enc_avx_gen4`/`aesni_gcm_dec_avx_gen4` and it worked, so I assume we can use the minimum padding.

**Comment by eugeneia** [comment]
@corsix @lukego I have added a few sanity tests to `lib.ipsec.aes_128_gcm`. Test results on a Intel(R) Xeon(R) CPU E5-1650 v3 @ 3.50GHz look as follows:

```
$ sudo ./snabb snsh -t lib.ipsec.aes_128_gcm
Test vector:    1
ciphertext  03 88 DA CE 60 B6 A3 92 F3 28 C2 B9 71 B2 FE 78
is          03 88 DA CE 60 B6 A3 92 F3 28 C2 B9 71 B2 FE 78
auth        58 E2 FC CE FA 7E 30 61 36 7F 1D 57 A4 E7 45 5A
is          58 E2 FC CE FA 7E 30 61 36 7F 1D 57 A4 E7 45 5A
plaintext   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
is          00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
auth        58 E2 FC CE FA 7E 30 61 36 7F 1D 57 A4 E7 45 5A
is          58 E2 FC CE FA 7E 30 61 36 7F 1D 57 A4 E7 45 5A
Test vector:    2
ciphertext  42 83 1E C2 21 77 74 24 4B 72 21 B7 84 D0 D4 9C E3 AA 21 2F 2C 02 A4 E0 35 C1 7E 23 29 AC A1 2E 21 D5 14 B2 54 66 93 1C 7D 8F 6A 5A AC 84 AA 05 1B A3 0B 39 6A 0A AC 97 3D 58 E0 91 47 3F 59 85
is          42 83 1E C2 21 77 74 24 4B 72 21 B7 84 D0 D4 9C E3 AA 21 2F 2C 02 A4 E0 35 C1 7E 23 29 AC A1 2E 21 D5 14 B2 54 66 93 1C 7D 8F 6A 5A AC 84 AA 05 1B A3 0B 39 6A 0A AC 97 3D 58 E0 91 47 3F 59 85
auth        32 47 18 4B 3C 4F 69 A4 4D BC D2 28 87 BB B4 18
is          32 47 18 4B 3C 4F 69 A4 4D BC D2 28 87 BB B4 18
plaintext   D9 31 32 25 F8 84 06 E5 A5 59 09 C5 AF F5 26 9A 86 A7 A9 53 15 34 F7 DA 2E 4C 30 3D 8A 31 8A 72 1C 3C 0C 95 95 68 09 53 2F CF 0E 24 49 A6 B5 25 B1 6A ED F5 AA 0D E6 57 BA 63 7B 39 1A AF D2 55
is          D9 31 32 25 F8 84 06 E5 A5 59 09 C5 AF F5 26 9A 86 A7 A9 53 15 34 F7 DA 2E 4C 30 3D 8A 31 8A 72 1C 3C 0C 95 95 68 09 53 2F CF 0E 24 49 A6 B5 25 B1 6A ED F5 AA 0D E6 57 BA 63 7B 39 1A AF D2 55
auth        32 47 18 4B 3C 4F 69 A4 4D BC D2 28 87 BB B4 18
is          32 47 18 4B 3C 4F 69 A4 4D BC D2 28 87 BB B4 18
Encrypted   100000000   bytes in    0.031974659010302   seconds
Decrypted   100000000   bytes in    0.03164627001388    seconds
```

Great work @corsix! Perfomance is as expected, 1.1 cycle per byte.

**Comment by corsix** [comment]
FWIW, while transliterating the assembly, I tested by cross-checking the DynASM and yasm implementations against each other. I wasn't happy with the test coverage until I covered every possible message length up to (IIRC) 2k bytes, which is to say that I don't believe your two test vectors provide very much test coverage.

**Comment by eugeneia** [comment]
@corsix For some reason, I can't get the encryption/decryption routines to “pick up” the “Additional Authentication Data”. E.g. the AAD doesn't impact the resulting ICV. I threw together a little test case: https://github.com/eugeneia/snabbswitch/blob/9fd0119f5a0bbb10c4b905c1b1e896e64d1cc142/src/lib/ipsec/aad_test.lua

I checked both the Intel routines and your transliteration and they exhibit the same behavior, so this really isn't your problem problem and I am guessing its some silly mistake in how I call the routines. Anyways, I can't make much sense of the assembler code so any hints would be highly appreciated.

**Comment by corsix** [comment]
Hmm, I'll investigate when I have time - which might be this Sunday, or might not be until next Saturday.

**Comment by corsix** [comment]
Try making (the contents of) `hash_subkey` non-zero before calling `aesni_gcm_precomp_avx_gen4`. Also note that `hash_subkey` should be 128 bits long rather than 128 bytes long.

For example, the [Intel test app](https://github.com/lukego/intel-ipsec/blob/master/code/LibTestApp/do_test.h#L289) does:

```
    DECLARE_ALIGNED(UINT8 hashSubKey[16], 16) = {15,14,13,12,11,10,9,8,
                                                 7,6,5,4,3,2,1,0};
```

If you look at the nice [GCM diagram from Wikipedia](https://en.wikipedia.org/wiki/Galois/Counter_Mode#/media/File:GCM-Galois_Counter_Mode.svg), `hash_subkey` corresponds to `H`. If `H` is zero, then every mult<sub>H</sub> step yields zero, and the auth tag ends up depending only on K and the IV (Counter 0). If `H` is any non-zero value, then the auth tag ends up also depending on the plaintext and the AAD.

**Comment by eugeneia** [comment]
@corsix D'oh. Thanks for the hint. I had mistakenly assumed that the Intel routines derive `hash_subkey` during `aesni_gcm_precomp`, but it seems we have to do that manually?

**Comment by corsix** [comment]
Indeed, `hash_subkey` is an input to `aesni_gcm_precomp` rather than an output (and it is in fact the sole input, with the other parameter being the sole output).

**Pull request closed without merging**